jmbitcoin/jmbitcoin/secp256k1_transaction.py

```diff
@@ -664,193 +175,103 @@ def mk_freeze_script(pub, locktime):
-    """
-    modtx = segwit_signature_form(deserialize(tx), i, redeem_script, amount,
-                                  hashcode, decoder_func=lambda x: x)
-    return ecdsa_tx_sign(modtx, pk, hashcode)
+    if not amount:
+        # p2pkh only supported here:
+        input_scriptPubKey = pubkey_to_p2pkh_script(pub)
+        sighash = SignatureHash(input_scriptPubKey, tx, i, hashcode)
+        try:
+            sig = ecdsa_raw_sign(sighash, priv, rawmsg=True) + bytes([hashcode])
+        except Exception as e:
+            return return_err(e)
+        tx.vin[i].scriptSig = CScript([sig, pub])
+        # Verify the signature worked.
+        try:
+            VerifyScript(tx.vin[i].scriptSig,
+                        input_scriptPubKey, tx, i, flags=flags)
+        except Exception as e:
+            return return_err(e)
+        return sig, "signing succeeded"
 
-def apply_p2wsh_multisignatures(tx, i, script, sigs):
-    """Sigs must be passed in as a list, and must be a
-    complete list for this multisig, in the same order
-    as the list of pubkeys when creating the scriptPubKey.
-    """
-    if isinstance(script, str):
-        script = binascii.unhexlify(script)
-    sigs = [binascii.unhexlify(x) if x[:2] == '30' else x for x in sigs]
-    if isinstance(tx, str):
-        return safe_hexlify(apply_p2wsh_multisignatures(
-            binascii.unhexlify(tx), i, script, sigs))
-    txobj = deserialize(tx)
-    txobj["ins"][i]["script"] = ""
-    txobj["ins"][i]["txinwitness"] = [None] + sigs + [script]
-    return serialize(txobj)
+    else:
+        # segwit case; we currently support p2wpkh native or under p2sh.
 
-def apply_multisignatures(*args):
-    # tx,i,script,sigs OR tx,i,script,sig1,sig2...,sig[n]
-    tx, i, script = args[0], int(args[1]), args[2]
-    sigs = args[3] if isinstance(args[3], list) else list(args[3:])
+        # see line 1256 of bitcointx.core.scripteval.py:
+        flags.add(SCRIPT_VERIFY_P2SH)
 
-    if isinstance(script, str):
-        script = binascii.unhexlify(script)
-    sigs = [binascii.unhexlify(x) if x[:2] == '30' else x for x in sigs]
-    if isinstance(tx, str):
-        return safe_hexlify(apply_multisignatures(
-            binascii.unhexlify(tx), i, script, sigs))
+        if native and native != "p2wpkh":
+            scriptCode = native
+            input_scriptPubKey = redeem_script_to_p2wsh_script(native)
+        else:
+            # this covers both p2wpkh and p2sh-p2wpkh case:
+            input_scriptPubKey = pubkey_to_p2wpkh_script(pub)
+            # only created for convenience access to scriptCode:
+            input_address = P2WPKHCoinAddress.from_scriptPubKey(
+                input_scriptPubKey)
+            # function name is misleading here; redeemScript only applies to p2sh.
+            scriptCode = input_address.to_redeemScript()
+
+        sighash = SignatureHash(scriptCode, tx, i, hashcode, amount=amount,
+                                sigversion=SIGVERSION_WITNESS_V0)
```

Comment:
A small thing, but you can also say `scriptCode.sighash(tx, i, hashcode, amount=amount, sigversion=SIGVERSION_WITNESS_V0)` and you won't need to import `SignatureHash`

Reply:
I've actually imported `*` from `core.script` (partly because I wanted the op codes), so it seems like I may as well leave this as-is?

Reply:
Of course, it is OK to leave as it is. I mentioned this just to inform about other option.